mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,307 @@
+            if (countTapes(func) >= 2) {
+                MultitapePrograms.push_back(func);
+            }
+        });
+
+        // Do nothing and exit for classical and single-tape programs
+        if (MultitapePrograms.empty()) {
+            return;
+        }
+
+        // 2. Parse the function into tapes
+        // total number of operation lists is number of tapes +2
+        // for classical pre and post processing
+        for (func::FuncOp func : MultitapePrograms) {
+            SmallVector<std::shared_ptr<SmallVector<Operation *>>> OpsEachTape(countTapes(func) + 2,
+                                                                               nullptr);
+
+            collectOperationsForEachTape(func, OpsEachTape);
+
+            // 3. Get the SSA values needed by the post processing (PP)
+            // These need to be returned by the tapes.
+            // Note that later tapes can also need values from earilier tapes.
+            SmallVector<Value> NecessaryValuesFromEarlierTapes;
+
+            // Go through all the operands of all the PP ops
+            // Find the ones that are not produced in PP itself
+            SmallVector<Operation *> PPOps = *(OpsEachTape.back());
+            collectNecessaryValuesFromEarlierTapes(PPOps, NecessaryValuesFromEarlierTapes);
+
+            // 4. Generate the functions for each tape
+            unsigned int NumTapes = countTapes(func);
+            SmallVector<FailureOr<func::FuncOp>> OutlinedFuncs;
+            for (unsigned int i = 0; i < NumTapes; i++) {
+                if (failed(createTapeFunction(OpsEachTape[OpsEachTape.size() - 2 - i],
+                                              NecessaryValuesFromEarlierTapes, builder,
+                                              OpsEachTape.size() - 3 - i, func, OutlinedFuncs))) {
+                    return signalPassFailure();
+                };
+            }
+
+            // OutlinedFuncs contains the tapes in reverse order (tape_2, tape_1, tape_0)
+            // Move them into the correct order
+            // Also, make the outlined functions keep the original's attributes
+            SmallVector<NamedAttribute> OutlinedFuncAttrs;
+            ArrayRef<NamedAttribute> FullOriginalFuncAttrs = func->getAttrs();
+            for (auto attr : FullOriginalFuncAttrs) {
+                StringRef attrname = attr.getName();
+                if ((attrname != "sym_name") && (attrname != "function_type")) {
+                    OutlinedFuncAttrs.push_back(attr);
+                }
+            }
+            for (auto outlined : OutlinedFuncs) {
+                func::FuncOp outlinedfunc = *outlined;
+                outlinedfunc->moveAfter(func);
+                outlinedfunc->setAttrs(OutlinedFuncAttrs);
+            }
+        }
+    } // runOnOperation()
+};
+
+std::unique_ptr<Pass> createSplitMultipleTapesPass()
+{
+    return std::make_unique<SplitMultipleTapesPass>();
+}
+
+} // namespace catalyst
```

Comment:
Strongly recommend to configure your editor to do this automatically (include final newline & strip trailing whitespace)
```suggestion
} // namespace catalyst

```